packages/android-lib/src/main/kotlin/com/outsystems/plugins/osgeolocation/controller/OSGLOCController.kt

```diff
@@ -0,0 +1,245 @@
+import com.outsystems.plugins.osgeolocation.model.OSGLOCException
+import com.outsystems.plugins.osgeolocation.model.OSGLOCLocationOptions
+import com.outsystems.plugins.osgeolocation.model.OSGLOCLocationResult
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.callbackFlow
+import kotlinx.coroutines.flow.first
+
+/**
+ * Entry point in OSGeolocationLib-Android
+ *
+ */
+class OSGLOCController(
+    fusedLocationClient: FusedLocationProviderClient,
+    activityLauncher: ActivityResultLauncher<IntentSenderRequest>,
+    private val helper: OSGLOCServiceHelper = OSGLOCServiceHelper(
+        fusedLocationClient,
+        activityLauncher
+    )
+) {
+    private lateinit var resolveLocationSettingsResultFlow: MutableSharedFlow<Result<Unit>>
+    private val locationCallbacks: MutableMap<String, LocationCallback> = mutableMapOf()
+    private val watchIdsBlacklist: MutableList<String> = mutableListOf()
+
+    /**
+     * Obtains the device's location using FusedLocationProviderClient.
+     * Tries to obtain the last retrieved location, and then gets a fresh one if necessary.
+     * @param activity the Android activity from which the location request is being triggered
+     * @param options OSLocationOptions object with the options to obtain the location with (e.g. timeout)
+     * @return Result<OSLocationResult> object with either the location or an exception to be handled by the caller
+     */
+    suspend fun getCurrentPosition(
+        activity: Activity,
+        options: OSGLOCLocationOptions
+    ): Result<OSGLOCLocationResult> {
+        try {
+            val checkResult: Result<Unit> =
+                checkLocationPreconditions(activity, options, isSingleLocationRequest = true)
+            return if (checkResult.isFailure) {
+                Result.failure(
+                    checkResult.exceptionOrNull() ?: NullPointerException()
+                )
+            } else {
+                val location = helper.getCurrentLocation(options)
+                return Result.success(location.toOSLocationResult())
+            }
+        } catch (exception: Exception) {
+            Log.d(LOG_TAG, "Error fetching location: ${exception.message}")
+            return Result.failure(exception)
+        }
+    }
+
+    /**
+     * Function to be called by the client after returning from the activity
+     * that is launched when resolving the ResolvableApiException in checkLocationSettings,
+     * that prompts the user to enable the location if it is disabled.
+     * @param resultCode to determine if the user enabled the location when prompted
+     */
+    suspend fun onResolvableExceptionResult(resultCode: Int) {
+        if (resultCode == Activity.RESULT_OK) {
+            resolveLocationSettingsResultFlow.emit(Result.success(Unit))
+        } else {
+            resolveLocationSettingsResultFlow.emit(
```

Comment:
maybe we could emit the `resolveLocationSettingsResultFlow.emit(...)` duplication?

Reply:
Refactored, can you check?